### DISCUSSION TITLE
Specifying build path with 4.0

### DISCUSSION THREAD

**Discussion by swilde** [fosskers/aura#860] [General]
I used to use `--build` argument with `-A` to specify the build directory where AUR packages are build and `--builduser` to set a specific user to build.  In 4.0 these options seem to be gone.  What is the correct way to emulate the previous behavior?

**Comment by fosskers** [comment]
Hi, thanks for the report. `--build` seems to have been overlooked - I will restore it for you tomorrow. In the meantime, you can set the build path in config. [See here](https://fosskers.github.io/aura/configuration.html#aur-package-building) for how to do that. Also consult [this page](https://fosskers.github.io/aura/aur.html) for how package building looks now. Mostly it's advice to not call `sudo` yourself.

As for the build user, packages are now built as your existing user, not through sudo-hacks as before, so I'm tempted to say you might not need `--builduser` anymore, but perhaps you have a unique usecase?

**Reply by swilde** [reply]
That seems to imply that you need to use aura as a user with sudoers rights so that aura is able to do system wide installation of packages and dependencies.  Quite frankly I do not like that very much, my way of doing things is to sun aura as root for installation and updates and rely on `--builduser` to drop privileges while building packages.

If this is no longer possible I will definitely miss this feature...

**Reply by fosskers** [reply]
[It's currently not possible to build AUR packages as root](https://github.com/fosskers/aura/issues/776), and the switch to the current behaviour actually solves a great many historical bugs. Perhaps this is something to address before releasing. Indeed it means that people trying to use Aura in simple root-only remote servers or Docker containers will have trouble.

I'm going to look into use of `pkgctl build` if the user is `root`. The core of the issue is that `makepkg` is not runnable as root, and there's no way around that. Note that `pkgctl build` is already usable for packages in an opt-in basis.

**Reply by fosskers** [reply]
`--build` has been restored.

**Reply by fosskers** [reply]
`pkgctl build` unfortunately can't be used as root either, as it calls `makepkg` internally.

**Reply by fosskers** [reply]
![dm](https://github.com/user-attachments/assets/2d33bb32-8a67-426f-bd2d-2fe3a5beba48)

Alright, it seems reasonable to utilise the `nobody` user in the invocation of `makepkg` if the user is detected to be `root`. This would enable the Arch ISO / server / Docker use-case.

**Reply by swilde** [reply]
Many thanks for your work and great support.  I'll give the reintroduced `--build` option a spin.

I'm fully aware that running `makepkg` as root is not permitted (and I think thats actually a good thing) but running aura as root, using the super user privileges to install dependencies and the resulting package seams reasonable.  For calling `makepkg` the privileges should/musty be dropped by switching to an unprivileged user, which to my understanding is basically the behavior up to this point.

Using the `nobody` user as an default in this case seems reasonable, if it would still be configurable it would be nice (I'm using a dedicated aura build user on my systems).  However an “if run as root, build as `nobody`” behavior would still meet my needs.

**Reply by fosskers** [reply]
Let's start with `nobody` and evolve from there.

**Reply by fosskers** [reply]
Building as `root` via `nobody` is now possible: https://github.com/fosskers/aura/releases/tag/v4.0.0-rc2

**Reply by swilde** [reply]
Thanks a lot.  I con confirm that the latest version can again be used as `root` and builds successfully as `nobody`.

**Reply by swilde** [reply]
FWIW, I just stumbled across another reason why I would prefer to be able to use a configurable user for building instead of `nobody`: some packages verify OpenPGP (gpg) signatures during the build process.  For this to work the users $HOME/.gnupg directory needs to be populated. However $HOME of `nobody` defaults to `/` which IMHO is not a good place for shenanigans of this type ...

**Reply by fosskers** [reply]
That seems to me to still be reasonable; it would just be a matter of setting the build user at the very beginning. If one is set by the user, then use it. Otherwise, use `nobody`. Note that setting `--build` doesn't work if the user is `root`, since the only directory that `nobody` has permission to write in is `/tmp`, but if a root user specified both `--build` and `--builduser` _and_ that user had permission to write into whatever `--build` is, then it would work.

 I imagine most people won't need to be this specific, but you're here asking for it, so there will likely be others.

**Reply by fosskers** [reply]
> packages verify OpenPGP (gpg) signatures during the build process.

For my own reference, can you name such a package so I can check its PKGBUILD?

**Reply by fosskers** [reply]
I just pushed a commit that restores `--builduser`.

**Reply by swilde** [reply]
Thanks a lot, I'll test it asap.
In the meantime: the package I last encountered with OpenPGP signitatur checking from aur is `python2`, but there are more for sure. Also I became aware of more times where the build user needs a writeable home: packages  compiling golang might want to write `$HOME/.cache`, and then there is haskell which writes `$HOME/.cargos` as `aura-git` for instance ... So thanks a lot for resorting `--builduser`!
(btw,  I created a build directory owned by `nobody` to use `--build` with the previous version)

**Reply by swilde** [reply]
I gave `--builduser` a spin and it seems to work as expected.  I'm very greatful for your great support!

**Reply by fosskers** [reply]
Good to know that there are packages that won't want to build cleanly as `root`. The `nobody` workaround is just that, a workaround, and power users can use `--builduser`, etc., as necessary.

**Comment by swilde** [comment]
I successfully tested that `--build` works again as expected.  Thanks.
